pkg/globalnet/controllers/ingress_endpoints_controller.go

```diff
@@ -0,0 +1,153 @@
+	var err error
+
+	_, gvr, err := util.ToUnstructuredResource(&submarinerv1.GlobalIngressIP{}, config.RestMapper)
+	if err != nil {
+		return nil, err
+	}
+
+	controller := &ingressEndpointsController{
+		baseSyncerController: newBaseSyncerController(),
+		svcName:              svc.Name,
+		namespace:            svc.Namespace,
+		ingressIPMap:         stringset.NewSynchronized(),
+	}
+
+	fieldSelector := fields.Set(map[string]string{"metadata.name": svc.Name}).AsSelector().String()
+
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:                "Ingress Endpoints syncer",
+		ResourceType:        &corev1.Endpoints{},
+		SourceClient:        config.SourceClient,
+		SourceNamespace:     svc.Namespace,
+		RestMapper:          config.RestMapper,
+		Federator:           federate.NewCreateFederator(config.SourceClient, config.RestMapper, corev1.NamespaceAll),
+		Scheme:              config.Scheme,
+		Transform:           controller.process,
+		SourceFieldSelector: fieldSelector,
+		ResourcesEquivalent: areEndpointsEqual,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := controller.Start(); err != nil {
+		return nil, err
+	}
+
+	ingressIPs := config.SourceClient.Resource(*gvr).Namespace(corev1.NamespaceAll)
+	controller.reconcile(ingressIPs, func(obj *unstructured.Unstructured) runtime.Object {
+		// TODO: Consider adding endpointsRef field to GlobalIngressIPSpec
+		endpointsName, exists, _ := unstructured.NestedString(obj.Object, "spec", "podRef", "name")
+		if exists {
+			return &corev1.Endpoints{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      endpointsName,
+					Namespace: obj.GetNamespace(),
+				},
+			}
+		}
+
+		return nil
+	})
+
+	klog.Infof("Created Endpoints controller for (%s/%s) with selector %q", svc.Namespace, svc.Name, fieldSelector)
+
+	return controller, nil
+}
+
+func (c *ingressEndpointsController) process(from runtime.Object, numRequeues int, op syncer.Operation) (runtime.Object, bool) {
+	endpoints := from.(*corev1.Endpoints)
+	key, _ := cache.MetaNamespaceKeyFunc(endpoints)
+
+	ingressIP := &submarinerv1.GlobalIngressIP{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("ep-%.59s", endpoints.Name),
+			Namespace: endpoints.Namespace,
+			Labels: map[string]string{
+				ServiceRefLabel: c.svcName,
+			},
+		},
+	}
+
+	if op == syncer.Delete {
+		c.ingressIPMap.Remove(ingressIP.Name)
+		klog.Infof("ingress Endpoints %s for service %s deleted", key, c.svcName)
+
+		return ingressIP, false
+	}
+
+	// TODO: consider other conditions
+	if c.ingressIPMap.Contains(ingressIP.Name) {
+		// Avoid assigning ingressIPs to endpoints that are not ready with an endpoint IP
+		return nil, false
+	}
+
+	klog.Infof("%q ingress Endpoints %s for service %s", op, key, c.svcName)
+
+	// TODO: consider handling multiple endpoints IPs
+	if len(endpoints.Subsets) == 0 || len(endpoints.Subsets[0].Addresses) == 0 {
```

Comment:
As per this implementation, if a user manually creates an endpoint with more than one ipAddress, we will only program rules for the first entry right?

Reply:
Yes. The current implementation only uses the first entry.

This commit meant to be just a quick prototype. I expect that multiple entries can be added, however I'm not sure how multiple `GlobalIngressIP`s can be handled by using admiral's syncer framework. Can it just return `GlobalIngressIPList`? (and then need to create a controller for it?)

Reply:
Note that this is still TODO.

Reply:
@sridhargaddam @anfredette @nerdalert 

> As per this implementation, if a user manually creates an endpoint with more than one ipAddress, we will only program rules for the first entry right?

The last commit adds support for multiple IPs in endpoints of headless service without selector. PTAL
(Note that first 2 commits will be merged through #1582 )

Current E2E test doesn't cover this specific case, but I've manually tested by below way and worked well:

<details>

1. Deploy the test environment

```
make deploy using=external-net,globalnet
```

2. Create headless service without selector with one endpoint and test connectivity and source IPs

- In cluster1, create headless service without selector with one endpoint
```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster1
 
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: test-vm
spec:
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
  clusterIP: None
EOF
 
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-vm
subsets:
  - addresses:
      - ip: 172.117.0.4
    ports:
      - port: 80
EOF
 
 
kubectl get svc,endpoints,globalingressip
NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   100.0.0.1    <none>        443/TCP   23m
service/test-vm      ClusterIP   None         <none>        80/TCP    27s
 
NAME                   ENDPOINTS         AGE
endpoints/kubernetes   172.18.0.5:6443   23m
endpoints/test-vm      172.117.0.4:80    27s
 
 
subctl export service -n default test-vm
```

- globalingressip is created after the service is exported
```
kubectl get svc,endpoints,globalingressip
NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   100.0.0.1    <none>        443/TCP   24m
service/test-vm      ClusterIP   None         <none>        80/TCP    55s
 
NAME                   ENDPOINTS         AGE
endpoints/kubernetes   172.18.0.5:6443   24m
endpoints/test-vm      172.117.0.4:80    55s
 
NAME                                                   IP
globalingressip.submariner.io/ep-test-vm-172.117.0.4   242.254.1.253
 
kubectl get -o yaml globalingressip.submariner.io/ep-test-vm-172.117.0.4
apiVersion: submariner.io/v1
kind: GlobalIngressIP
metadata:
  annotations:
    submariner.io/headless-svc-endpoints-ip: 172.117.0.4
  creationTimestamp: "2022-03-08T22:48:58Z"
  generation: 2
  labels:
    submariner-io/originatingNamespace: default
    submariner.io/serviceRef: test-vm
  name: ep-test-vm-172.117.0.4
  namespace: default
  resourceVersion: "5297"
  uid: 46ed1f13-e65c-4240-8374-0935313d134b
spec:
  serviceRef:
    name: test-vm
  target: HeadlessServiceEndpoints
status:
  allocatedIP: 242.254.1.253
  conditions:
  - lastTransitionTime: "2022-03-08T22:48:58Z"
    message: Allocated global IP
    reason: Success
    status: "True"
    type: Allocated
``` 
 
- In cluster2, create statefulset, service without selector and export the service.
```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster2
 
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Service
metadata:
  name: nginx-ss
  labels:
    app.kubernetes.io/instance: nginx-ss
    app.kubernetes.io/name: nginx-ss
spec:
  ports:
  - port: 80
    name: web
  clusterIP: None
  selector:
    app.kubernetes.io/instance: nginx-ss
    app.kubernetes.io/name: nginx-ss
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: web
spec:
  serviceName: "nginx-ss"
  replicas: 2
  selector:
    matchLabels:
        app.kubernetes.io/instance: nginx-ss
        app.kubernetes.io/name: nginx-ss
  template:
    metadata:
      labels:
        app.kubernetes.io/instance: nginx-ss
        app.kubernetes.io/name: nginx-ss
    spec:
      containers:
      - name: nginx-ss
        image: k8s.gcr.io/nginx-slim:0.8
        ports:
        - containerPort: 80
          name: web
EOF
 
subctl export service -n default nginx-ss
```

- Test access from pod to external
```
kubectl get globalingressip
NAME        IP
pod-web-0   242.254.2.252
pod-web-1   242.254.2.253
 
kubectl exec -it web-0 -- curl -m 10 242.254.1.253
 
docker logs ext-app 2>&1 | tail -n 1
242.254.2.252 - - [08/Mar/2022 22:56:55] "GET / HTTP/1.1" 200 -
 
kubectl exec -it web-1 -- curl -m 10 242.254.1.253
 
docker logs ext-app 2>&1 | tail -n 1
242.254.2.253 - - [08/Mar/2022 22:57:40] "GET / HTTP/1.1" 200 -
```

- Test access from external to pod
```
docker exec -it ext-app curl -m 10 242.254.2.252
 
kubectl logs web-0
242.254.1.253 - - [08/Mar/2022:22:59:11 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
 
 
docker exec -it ext-app curl -m 10 242.254.2.253
kubectl logs web-1
242.254.1.253 - - [08/Mar/2022:23:00:12 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
```

3. Create another test container, add it to endpoints, and test connectivity and source IPs

- Create another test container
```
docker run --cap-add=NET_ADMIN --name ext-app2 --network pseudo-ext \
-d registry.access.redhat.com/ubi7/ubi python -m SimpleHTTPServer 80
 
docker exec -i ext-app2 yum install -y iproute
docker exec -i ext-app2 ip r add 242.254.1.0/24 via 172.117.0.3
docker exec -i ext-app2 ip r add 242.254.2.0/24 via 172.117.0.3
```

- Add it to endpoints (Then new ingressglobalip ep-test-vm-172.117.0.5 is added)
```
docker inspect ext-app2 -f '{{(index .NetworkSettings.Networks "pseudo-ext").IPAddress}}'
172.117.0.5
 
export KUBECONFIG=output/kubeconfigs/kind-config-cluster1
 
cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-vm
subsets:
  - addresses:
      - hostname: ext-app
        ip: 172.117.0.4
      - hostname: ext-app2
        ip: 172.117.0.5
    ports:
      - port: 80
EOF
 
kubectl get svc,endpoints,globalingressip
NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   100.0.0.1    <none>        443/TCP   52m
service/test-vm      ClusterIP   None         <none>        80/TCP    28m
 
NAME                   ENDPOINTS                       AGE
endpoints/kubernetes   172.18.0.5:6443                 52m
endpoints/test-vm      172.117.0.4:80,172.117.0.5:80   28m
 
NAME                                                   IP
globalingressip.submariner.io/ep-test-vm-172.117.0.4   242.254.1.253
globalingressip.submariner.io/ep-test-vm-172.117.0.5   242.254.1.252
```
 
- Test access from pod to external
```
export KUBECONFIG=output/kubeconfigs/kind-config-cluster2
 
kubectl exec -it web-0 -- curl -m 10 242.254.1.252
 
docker logs ext-app2 2>&1 | tail -n 1
242.254.2.252 - - [08/Mar/2022 23:19:05] "GET / HTTP/1.1" 200 -
 
kubectl exec -it web-1 -- curl -m 10 242.254.1.252
 
docker logs ext-app2 2>&1 | tail -n 1
242.254.2.253 - - [08/Mar/2022 23:19:25] "GET / HTTP/1.1" 200 -
```

- Test access from external to pod
```
docker exec -it ext-app2 curl -m 10 242.254.2.252
 
kubectl logs web-0
242.254.1.253 - - [08/Mar/2022:22:59:11 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
242.254.1.252 - - [08/Mar/2022:23:20:31 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
 
docker exec -it ext-app2 curl -m 10 242.254.2.253
kubectl logs web-1
242.254.1.253 - - [08/Mar/2022:23:00:12 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
242.254.1.252 - - [08/Mar/2022:23:20:42 +0000] "GET / HTTP/1.1" 200 612 "-" "curl/7.29.0"
```

</details>